src/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -282,262 +280,530 @@ internal static ulong ToUInt64(object value)
-            else
-                throw parseResult.GetEnumParseException();
+            Debug.Assert(typeof(TResult) == enumType || typeof(TResult) == typeof(object));
+            Debug.Assert(enumType.GetEnumUnderlyingType() == typeof(uint));
+
+            bool failureIsOverflow = false;
+            if (StartsNumber(value[0]) && Number.TryParseUInt32IntegerStyle(value, NumberStyles.AllowLeadingSign | NumberStyles.AllowTrailingWhite, CultureInfo.InvariantCulture.NumberFormat, out uint uintResult, ref failureIsOverflow))
+            {
+                enumResult = RuntimeHelpers.IsReferenceOrContainsReferences<TResult>() ? (TResult)InternalBoxEnum(enumType, uintResult) : Unsafe.As<uint, TResult>(ref uintResult);
+                return true;
+            }
+
+            if (failureIsOverflow)
+            {
+                if (throwOnFailure) throw new OverflowException(SR.Overflow_UInt32);
+            }
+            else if (TryParseByName(enumType, originalValueString, value, ignoreCase, throwOnFailure, out ulong ulongResult))
+            {
+                uintResult = (uint)ulongResult;
+                enumResult = RuntimeHelpers.IsReferenceOrContainsReferences<TResult>() ? (TResult)InternalBoxEnum(enumType, uintResult) : Unsafe.As<uint, TResult>(ref uintResult);
+                return true;
+            }
+
+            enumResult = default;
+            return false;
         }
 
-        private static bool TryParseEnum(Type enumType, string value, bool ignoreCase, ref EnumResult parseResult)
+        private static bool TryParseInt64Enum<TResult>(RuntimeType enumType, string originalValueString, ReadOnlySpan<char> value, bool ignoreCase, bool throwOnFailure, out TResult enumResult)
         {
-            if (enumType == null)
-                throw new ArgumentNullException(nameof(enumType));
+            Debug.Assert(typeof(TResult) == enumType || typeof(TResult) == typeof(object));
+            Debug.Assert(enumType.GetEnumUnderlyingType() == typeof(long));
 
-            RuntimeType rtType = enumType as RuntimeType;
-            if (rtType == null)
-                throw new ArgumentException(SR.Arg_MustBeType, nameof(enumType));
+            bool failureIsOverflow = false;
+            if (StartsNumber(value[0]) && Number.TryParseInt64IntegerStyle(value, NumberStyles.AllowLeadingSign | NumberStyles.AllowTrailingWhite, CultureInfo.InvariantCulture.NumberFormat, out long longResult, ref failureIsOverflow))
+            {
+                enumResult = RuntimeHelpers.IsReferenceOrContainsReferences<TResult>() ? (TResult)InternalBoxEnum(enumType, longResult) : Unsafe.As<long, TResult>(ref longResult);
+                return true;
+            }
 
-            if (!enumType.IsEnum)
-                throw new ArgumentException(SR.Arg_MustBeEnum, nameof(enumType));
+            if (failureIsOverflow)
+            {
+                if (throwOnFailure) throw new OverflowException(SR.Overflow_Int64);
```

Comment:
nit: brace brackets here since it is in a nested scope.

Reply:
Done